### PR TITLE
[UPnP] Add extensions to libplatinum mimetypes

### LINKED ISSUE #22994: Wrong mime type for m2ts video in Kodi UPNP server
### Describe the bug
Here is a clear and concise description of what the problem is:

Wrong mime type for m2ts video in Kodi UPNP server: video/m2ts.
Correct mime type: video/vnd.dlna.mpeg-tts.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

I want to play m2ts video by BD player (Sony UBR-X700) from Kodi UPNP server.

## Actual Behavior

BD player (Sony UBR-X700) does not see this video but sees and plays mkv video from Kodi UPNP server.

## Possible Fix

Correct mime type for m2ts video in Kodi UPNP server: video/vnd.dlna.mpeg-tts.

Media Type 'video/vnd.dlna.mpeg-tts' Details:
This media type identifies audiovisual streams multiplexed using an extended MPEG-2 Transport Stream packet format.

I made the tests with the same m2ts video from Kodi, Plex and Emby UPNP servers.

Kodi UPNP server XML:

< res duration="0:47:12.000" resolution="1920x1080" protocolInfo="http-get:*:video/m2ts:*" > http : // 192.168.1.5:1782/%25/d29d7c1164e2eb9852f723f494204d15/file.m2ts < /res >

BD player does not see this video.

Plex UPNP server XML (the same from Emby):

< res duration="0:47:12.000" size="1389533184" resolution="1920x1080" bitrate="490500" nrAudioChannels="2" protocolInfo="http-get:*:video/vnd.dlna.mpeg-tts: DLNA.ORG_OP=01;DLNA.ORG_CI=0;DLNA.ORG_FLAGS=01500000000000000000000000000000" > http : // 192.168.1.5:32469/object/6ff06b80f68469b60739/file.ts < /res >

BD player sees and plays this video

Only Kodi UPNP client can see m2ts videos from Kodi UPNP server.

### To Reproduce

1. Add m2ts video to Library.
2. Try look and play this m2ts video by some UPNP player (not Kodi)

### Debuglog

The debuglog can be found here:
The problem is on non-Kodi players. No debuglog.

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [x] Linux
 - [ ] macOS
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name:
 - Kodi version:
Ubuntu 22.04, Linux desktop 5.15.0-57-generic # 63-Ubuntu SMP Thu Nov 24 13:43:17 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
Kodi (20.1 (20.1.0) Git:20230312-289ec664e3). Platform: Linux x86 64-bit

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by enen92**
## Description
This PR adds a new map that allows kodi to define mimetype extensions (or to override existing ones) to those already defined by libplatinum. This map takes precedence over all the others.
Libplatinum lacks the mapping for m2ts which leads external clients not to recognise the item.

## Motivation and context
Fixes https://github.com/xbmc/xbmc/issues/22994

## How has this been tested?
Kodi as server, philips tv default file browser as client

## What is the effect on users?
Better support for upnp

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [x] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

File: xbmc/network/upnp/UPnPInternal.cpp [hunk 1]
```diff
@@ -141,12 +147,28 @@ GetMimeType(const CFileItem& item,
...
+      }
+    }
+
+    /* Give priority to Platinum mime types as they are defined to map extension to DLNA compliant mime types
+       or custom types according to context (who asked for it) */
+    if (!ext.IsEmpty() && mime.IsEmpty())
+    {
+      mime = PLT_MimeType::GetMimeTypeFromExtension(ext, context);
+      if (mime == "application/octet-stream")
+        mime = "";
```
Comment (fuzzard) [inline xbmc/network/upnp/UPnPInternal.cpp:171]: Worthwhile just dropping the isempty check and putting this block inside the previous isempty check after the for loop?

**Review by fuzzard — APPROVED** [review]
Seems like a clean way to extend an otherwise forgotten lib

File: xbmc/network/upnp/UPnPInternal.cpp [hunk 2]
```diff
@@ -141,12 +147,28 @@ GetMimeType(const CFileItem& item,
...
-       as it is defined to map extension to DLNA compliant mime type
-       or custom according to context (who asked for it) */
-    if (!ext.IsEmpty()) {
-        mime = PLT_MimeType::GetMimeTypeFromExtension(ext, context);
-        if (mime == "application/octet-stream") mime = "";
+    /* We look first to our extensions/overrides of libplatinum mimetypes. If not found, fallback to
+     Platinum definitions.  */
+    if (!ext.IsEmpty())
+    {
+      for (const NPT_HttpFileRequestHandler_DefaultFileTypeMapEntry& mimeTypeEntry :
```
Comment (Montellese) [inline xbmc/network/upnp/UPnPInternal.cpp:154]: Maybe use `std::find_if()` instead of writing the loop yourself.

**Review by Montellese — CHANGES_REQUESTED** [review]
Looks good apart from the proposed changes.

**Comment by enen92** [comment]
thanks for the review, should be fixed in the last iteration.

File: xbmc/network/upnp/UPnPInternal.cpp [hunk 3]
```diff
@@ -141,15 +147,34 @@ GetMimeType(const CFileItem& item,
...
-    /* We always use Platinum mime type first
-       as it is defined to map extension to DLNA compliant mime type
-       or custom according to context (who asked for it) */
-    if (!ext.IsEmpty()) {
+    if (!ext.IsEmpty())
+    {
+      /* We look first to our extensions/overrides of libplatinum mimetypes. If not found, fallback to
+         Platinum definitions.
+      */
+      auto KodiOverrideMimeType = std::find_if(
```
Comment (Montellese) [inline xbmc/network/upnp/UPnPInternal.cpp:155]: Can be `const`. And shouldn't all local variables start with a lower case letter?

File: xbmc/network/upnp/UPnPInternal.cpp [hunk 4]
```diff
@@ -141,15 +147,34 @@ GetMimeType(const CFileItem& item,
...
+          std::begin(KodiPlatinumMimeTypeExtensions), std::end(KodiPlatinumMimeTypeExtensions),
+          [&](const NPT_HttpFileRequestHandler_DefaultFileTypeMapEntry& mimeTypeEntry) {
```
Comment (Montellese) [inline xbmc/network/upnp/UPnPInternal.cpp:157]:
```suggestion
          [&](const auto& mimeTypeEntry) {
```

**Review by Montellese — APPROVED** [review]
Looks good apart from the suggestions.

**Review by garbear — APPROVED** [review]
LGTM

**Pull request merged**